File: Algorithms/977. Squares of a Sorted Array/kotlin/solution.two-pointers.kt
class Solution {
    fun sortedSquares(nums: IntArray): IntArray {
        val n = nums.size
        val result = IntArray(n)
        var left = 0
        var right = n - 1
        var idx = n - 1

        while (left <= right) {
            val leftSquare = nums[left] * nums[left]
            val rightSquare = nums[right] * nums[right]

            if (leftSquare <= rightSquare) {
                result[idx] = rightSquare
                right--
            } else {
                result[idx] = leftSquare
                left++
            }
            idx--
        }

        return result
    }
}

File: Algorithms/977. Squares of a Sorted Array/rust/solution.two-pointers.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut result = vec![0; nums.len()];
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        for i in (0..nums.len()).rev() {
            let mut square = 0;
            let left_square = nums[left as usize] * nums[left as usize];
            let right_square = nums[right as usize] * nums[right as usize];
            if left_square > right_square {
                result[i] = left_square;
                left += 1;
            } else {
                result[i] = right_square;
                right -= 1;
            }
        }

        result
    }
}